### DISCUSSION TITLE
chap 8-1 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#25] [General]
chap 8 - memory management

**Comment by sihyung92** [comment]
# 느낀 점
- 락도 그렇고 데드락도 그렇고 메모리도 그렇고, 현재는 사용되지 않는 부분에 대한 이야기의 비중이 크다. 나중에 알게 되었을 때 살짝 허무한 감이 있다.
- [무어의 법칙](https://ko.wikipedia.org/wiki/%EB%AC%B4%EC%96%B4%EC%9D%98_%EB%B2%95%EC%B9%99) 에 의해 램이 끊임없이 커지고 있는데도 (지금만 해도 32기가바이트 램을 쓰고있다) 프로세스가 점점 무거워지니 맨날 원점인거 같다... 요 귀한 램을 잘 쪼개 쓰려는 노력이 있기에 지금의 성능도 보장받는거겠지 싶다.
- segmentation에 대한 설명이 부실한거 같다. paged segmentation도 그냥 넘어갔으니 나중에 나오겠지?

# 문제
다음 문제들의 틀린 부분을 짚어주세요!

- 사용자 프로세스를 할당할 때 가변분할 방식(Variable Partition allocation)을 활용하면 메모리 단편화(조각 발생)문제를 해결할 수 있다. 
- 주소 동적 바인딩을 하면, 프로세스가 자신의 프로세스가 포함되지 않은 다른 메모리 주소의 data나 instruction을 호출할 위험이 있다.
- 페이징 기법을 활용할 때, 논리 주소의 모든 페이지의 정보를 페이지 테이블로 만들어 메모리에 보관한다.

**Reply by 1chz** [reply]
- 사용자 프로세스를 할당할 때 가변분할 방식(Variable Partition allocation)을 활용하면 메모리 단편화(조각 발생)문제를 해결할 수 **없다.**
- 주소 동적 바인딩을 하면, 프로세스가 자신의 프로세스가 포함되지 않은 다른 메모리 주소의 data나 instruction을 호출할 위험이 **없다.**
- 페이징 기법을 활용할 때, 논리 주소의 모든 페이지의 정보를 페이지 테이블로 만들어 메모리에 보관한다. (?)

**Reply by inseo24** [reply]
- 메모리 단편화는 문제는 여전히 발생
- 실행 중엔 변경되지 않음
- 맞는듯?

**Reply by sihyung92** [reply]
1. 정적분할에서 발생하는 것 보단 낫지만, 여전히 프로세스가 할당되고 해제되는 과정에서 외부단편화가 발생하게 된다 (Hole 문제)
2. limit register 값보다 큰 범위를 찌르게 되면 trap을 발생시키는 방식으로 다른 프로세스를 찌르지 못하게 보호한다.
3. 2중 테이블 구조에서, 모든 페이지가 아니라 활용하는 페이지만 기록하는 방식으로 활용한다.

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- logical address를 physical address로 매핑해주는 하드웨어 장치가 있는 걸 알게 됐다.
- MMU에서 relocation register(base register)와 limit register의 역할이 각각 분리되어 있는 점이 재밌었다.
- swap in/out 부분에서 오랜만에(?) 스케줄러 얘기가 나와서 반가웠다. 스터디 후에도 아마 운영체제 공부를 개인적으로 할 예정인데 이 부분과 연결해서 공부하면 더 좋을 거 같다.
- 강의 마지막 부분에 logical address의 크기에 대해서 정확히 이해하지 못했다. 


<hr />
<h3>문제</h3>

Q. 2단계 페이징 테이블이 속도상, 공간상 큰 이점을 주지 못함에도 사용하는 이유가 무엇일까요?

**Reply by 1chz** [reply]
- 현대의 컴퓨터는 하드웨어의 발달로 메모리를 크게 잡아먹는 프로그램을 지원
  - 예를 들어 32bit 주소 체계를 사용하는 컴퓨터에서는 4GB의 주소 공간을 갖는 프로그램을 지원
    - 페이지 사이즈가 4K라면, 한 프로세스당 페이지 테이블을 위해 1M 크기의 페이지 테이블 메모리 공간이 필요
    - 그러나 대부분의 프로그램은 4G의 주소 공간 중 지극히 일부분만 사용하므로 페이지 테이블 공간이 심하게 낭비 됨
    - 이러한 공간 낭비를 막기 위해 페이지 테이블 자체를 페이지로 구성하는 다단계 페이징 기법을 사용

**Reply by sihyung92** [reply]
2단계 페이징 테이블은 1단계보다 한 번 더 메모리 엑세스를 발생시킴으로써 속도상에는 이점이 없으나, 사용하지 않는 가상 메모리 공간에 대해서 outer table에서 null 값을 할당 함으로써 공간상 이점을 볼 수 있습니닷

**Comment by 1chz** [comment]
# 느낀점

---

- 컴퓨터 과학에서는 `Indirection` 이 은탄환과 같다는 얘기를 들은적이 있는데, 중간 계층을 두어 간접참조를 하면 정말 대부분의 문제를 해결할 수 있다고 느꼈다. 예를 들자면 다음과 같은 사례들이 있다.
  - 자바에서는 데이터베이스에 독립적이기 위해 JPA라는 `추상화 된 데이터베이스(Persistence Context)` API를 제공한다
  - 자바는 운영체제에 독립적이기 위해(플랫폼 독립성) JVM이라는 `추상화 된 운영체제`를 사용한다
  - 운영체제는 모든 사용자 프로세스가 자신의 시작 주소를 0번지인걸로 알게끔 하기 위해 MMU라는 간접참조 계층을 둔다
- 이제는 역사의 뒤안길로 사라진 개념들이 많이 나와 혼란스러운 느낌이 들긴 하지만, 비교를 위해 설명한다고 생각하니 그럴 수 있겠다는 생각이 들었다
- 메모리 단편화를 줄이는 작업을 하드디스크에서하면 디스크 조각모음이 아닌가? 라는 생각이 들었다
- 자바 고급 서적들에서 설명하던 내용들 중 이해되지 않았던 많은 내용들이 이번에 많이 이해된 것 같다
  - 페이징 어쩌고 하던 이야기, TLB, GC 등등

<img width="501" alt="스크린샷 2022-03-06 오후 7 05 00" src="https://user-images.githubusercontent.com/71188307/156918499-997247ff-a80c-4baa-bea5-b3030a4ac235.png">
<img width="488" alt="스크린샷 2022-03-06 오후 7 05 14" src="https://user-images.githubusercontent.com/71188307/156918501-af488354-fb3d-42de-968b-ddcf8d995f31.png">


# 문제

---

- 외부 단편화와 내부 단편화에 대해 서술하시오

**Reply by inseo24** [reply]
외부단편화 - 올리려는 프로그램 크기보다 메모리가 더 작아서 사용이 안되는 경우
내부단편화 - 프로그램 크기가 메모리보다 작아서 그 안에서 메모리가 남는 상황

**Reply by sihyung92** [reply]
메모리 단편화는 메모리에 프로세스를 할당하는 과정에서 발생하는 공간 낭비를 일컫는다. 고정 분할 방식에서 어떠한 프로세스도 진입할 수 없을 정도로 작은 조각인 외부 단편화가 발생하며, 가변 분할 방식과 페이징 기법에서 프로세스가 메모리의 공간을 할당 받았을 때 프로세스가 활용할 것보다 많은 메모리를 할당받은 내부 단편화가 일어나게 된다.
